### DISCUSSION TITLE
How to test HTML validation popup?  / Documentation for mocking typescript onSubmit()

### DISCUSSION THREAD

**Discussion by AkselAllas** [react-hook-form/react-hook-form#3151] [Q&A] [answered]
There is a nice testing section documentation, but I have a problem where I want to test this HTML validation:
![image](https://user-images.githubusercontent.com/26136082/95718026-87dcc980-0c76-11eb-9d10-64c749171d9b.png)

My solution would be to test whether handleSubmit() or onSubmit() has been called, using the solution here: https://github.com/react-hook-form/react-hook-form/issues/532

I have the following code:

```
const TestComponent = (
  onSubmit: any
) => {
  const reactHookFormMethods = useForm<MyFormType>();
  const { handleSubmit } = reactHookFormMethods;
  return (
    <FormProvider {...reactHookFormMethods}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <FormToTest />
        <Button type="submit">
          Submit
        </Button>
      </form>
    </FormProvider>
  );
};
```

```
  it('should not allow submitting negative number', async () => {
    const onSubmit = jest.fn();
    render(
      <Provider store={store}>
        <TestComponent onSubmit={onSubmit} />
      </Provider>
    );
    const input = screen.getByLabelText('input label');
     fireEvent.input(input, {
      target: {
        value: '-12',
      },
     });
    await act(async () => {
      fireEvent.submit(screen.getByRole('button', { name: /Submit/i }));
    });
    expect(onSubmit).toBeCalled();
  });
});
```
But with this I get:
`TypeError: onValid is not a function`

**Comment by bluebill1049** [comment]
can you reproduce this in a codesandbox with test?

**Reply by bluebill1049** [reply]
maybe `novalidate` at form?

**Reply by AkselAllas** [reply]
novalidate didn't help my case :(

**Comment by AkselAllas** [comment]
I tried making a codesandbox, but codesandbox has it's own bug i think: https://github.com/testing-library/dom-testing-library-template/issues/4

https://codesandbox.io/s/eager-albattani-f7zw4?file=/src/TestComponent.test.tsx

**Comment by AkselAllas** [comment]
Because codesandbox test failed, I made a repository of my own to demo this.

https://github.com/AkselAllas/RHF-typescript-test-problem

**Reply by bluebill1049** [reply]
kk will take a look at it.

**Comment by bluebill1049** [comment]
<img width="845" alt="Screen Shot 2020-10-13 at 12 47 02 pm" src="https://user-images.githubusercontent.com/10513364/95805894-3275ea80-0d52-11eb-98e7-12e58e467a1e.png">

this is the problem i am seeing at my end: https://react-hook-form.com/faqs#TestingfailedduetoMutationObserver

**Reply by AkselAllas** [reply]
I'm sorry for the mishap. I failed to commit the correct package. At first I tried using react-mutation-observer. Now it works.

You can now pull the latest version of the repo and see the `onValid is not a function` error

**Comment by bluebill1049** [comment, marked as answer]
<img width="232" alt="Screen Shot 2020-10-13 at 4 46 30 pm" src="https://user-images.githubusercontent.com/10513364/95820344-ab396e80-0d73-11eb-99f7-b2a59521d603.png">
please double check your props.

**Reply by AkselAllas** [reply]
I have tried a bunch of things:

- `any`
- `Function`
- `SubmitHandler<Function>`
- `SubmitHandler<Function> => any`
- `<MyForm>(onValid: SubmitHandler<MyForm>, onInvalid?: SubmitErrorHandler<MyForm> | undefined) => (e?: React.BaseSyntheticEvent<object, any, any> | undefined)`
- `SubmitHandler<MyForm>`

This is my main problem. I don't quite understand the typing :(

**Reply by bluebill1049** [reply]
It should be ({ onSubmit }) it's a prop right?

**Reply by AkselAllas** [reply]
Holy hell. That was the problem indeed. Thank you!

At least this made me do a bunch of research into generics, type assertions and even RHT source code.
